### ISSUE TITLE
Device Code Grant Type

### ISSUE THREAD

**Issue by jcampbell05** [oauthlib/oauthlib#626]
**Describe the feature**

In the latest spec there is a device code grant type https://www.oauth.com/oauth2-servers/device-flow/token-request/

This is useful for authenticating embedded devices. Right now this library lacks this feature.

**Comment by JonathanHuot** [comment]
The Spec is still in draft https://tools.ietf.org/html/draft-ietf-oauth-device-flow-13, but it will be great if anyone start to implement it in oauthlib. PRs are welcomed

**Comment by jcampbell05** [comment]
@JonathanHuot is there a good documentation on where contributors can begin to add this device ?

**Comment by thedrow** [comment]
You can ask on our Gitter channel.

**Comment by JonathanHuot** [comment]
As thedrow mention it, we're very reactive in Gitter. You can also have a look on the other grants and see how they are currently implemented. Also note that we are differenting the _client_ code from the _server_ code, but we have sometime common code to both client and server.
If you are looking into the client part, ideally, you should also propose a PR to `requests-oauthlib` which is the main "client" downstream of oauthlib.

You can also have a look on our Contributing page https://oauthlib.readthedocs.io/en/latest/contributing.html for more generic stuff.

**Comment by jcampbell05** [comment]
I'm more interested in the server side as I need to implement this for a project I'm working on. Atm I've done it using my custom code but I very much want to do it the correct way.

Will jump on gitter to discuss :)

**Comment by duaneking** [comment]
I cant help but feel like leaving the client side out is a bad idea.

**Comment by jcampbell05** [comment]
I can build the client side but the device code server itself is still alot a work

**Comment by duaneking** [comment]
I'm just thinking of the architecture needed for this in the context of the VERY needed JWT support.

**Comment by JonathanHuot** [comment]
@duaneking, full JWT support may have architecture changes but it should not prevent @jcampbell05 of completing his work. The Device Flow has nothing to do with JWT and should not be conflicting with the JWT new architecture.

About client's side, if it's implemented during the first PR, that's would be great I confess. Also, it helps to test the Device Flow implementation itself.

**Comment by duaneking** [comment]
My biggest push back was on having features be half done; If you are adding a client then that resolves this.  Thank You.

**Comment by jcampbell05** [comment]
I don't have bandwidth to finish this right now - so hopefully someone can pick this up or perhaps I can resume later on

**Comment by wiliamsouza** [comment]
No more in draft https://tools.ietf.org/html/rfc8628

**Comment by oliamb** [comment]
> So if we could identify ways of splitting it up into smaller chunks that might be manageable.

I can open a PR on Section 3.1 and 3.2 of the spec. Would that help? I am new to the codebase but I can try to stick to what exists already and maybe get some guidance on the expected changes afterward? That would cover the Device Authorization request and response.

PR #640 is still linked to the Draft specification, so the comments are a bit outdated. Is it ok to use if as an inspiration and start fresh?

**Comment by jcampbell05** [comment]
Yeah no problem, that's what I was hoping would happen. I'll have a look at the PR 😀

**Comment by brianhelba** [comment]
@duzumaki @thedrow Is this issue fully resolved by #881, or is there additional work beyond
> Implements Section 3.1 and Section 3.2 of [RFC8628](https://www.rfc-editor.org/rfc/rfc8628#section-3.1)

to satisfy this?

**Comment by duzumaki** [comment]
@brianhelba it seems so. enough to close this issue and if there's anything missing it would be minor and a follow up pr can be merged

I did have to merge [this](https://github.com/oauthlib/oauthlib/pull/885)

I'm working on a pr for `django-oauth-toolkit` that is using this for a custom auth server I'm working on

**Comment by duzumaki** [comment]
@brianhelba enough to close this issue and if there's anything missing it would be minor and a follow up pr can be merged

I did have to merge [this](https://github.com/oauthlib/oauthlib/pull/885)

I'm working on a pr for `django-oauth-toolkit` that is using this for a custom auth server I'm working on

**Comment by duzumaki** [comment]
@brianhelba section 3.4 and 3.5 need to be implemented. I'm working on it currently

**Comment by duzumaki** [comment]
@auvipy Would you be able to review this please
https://github.com/oauthlib/oauthlib/pull/891

Found 2 things while working on 
https://github.com/jazzband/django-oauth-toolkit/pull/1539
